ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java

```diff
@@ -1099,21 +1107,43 @@ private void drawRectangularBackgroundWithBorders(Canvas canvas) {
           int bottom = bounds.bottom;
 
           mPaint.setColor(fastBorderColor);
+          mPaint.setStyle(Paint.Style.STROKE);
           if (borderLeft > 0) {
             int leftInset = left + borderLeft;
-            canvas.drawRect(left, top, leftInset, bottom - borderBottom, mPaint);
+            Path path = new Path();
```

Comment:
It is generally recommended against creating new instances in functions called in `onDraw`. You may want to consider creating the `Path` where `mPaint` is created and reset it before using it here.

Reply:
@benoitdion I am just putting this pull request out there as as Quickfix for https://github.com/facebook/react-native/issues/24224 and 25+ issues with this bug https://github.com/facebook/react-native/issues/24224 https://github.com/facebook/react-native/issues/28695 https://github.com/facebook/react-native/issues/23651 https://github.com/facebook/react-native/issues/23475 https://github.com/facebook/react-native/issues/22256 https://github.com/facebook/react-native/issues/22226 https://github.com/facebook/react-native/issues/19234 https://github.com/facebook/react-native/issues/18285 https://github.com/facebook/react-native/issues/17344 https://github.com/facebook/react-native/issues/17343 https://github.com/facebook/react-native/issues/17251 https://github.com/facebook/react-native/issues/12817 https://github.com/facebook/react-native/issues/12403 https://github.com/facebook/react-native/issues/11042 https://github.com/facebook/react-native/issues/9343 https://github.com/facebook/react-native/issues/8236 https://github.com/facebook/react-native/issues/8105 https://github.com/facebook/react-native/issues/7838 https://github.com/facebook/react-native/issues/6721 https://github.com/facebook/react-native/issues/5411 https://github.com/facebook/react-native/issues/3159 https://github.com/facebook/react-native/issues/2335 https://github.com/facebook/react-native/issues/840 https://github.com/facebook/react-native/issues/27133 https://github.com/facebook/react-native/issues/28695 

All those issues have been closed, but actually never fixed. 

I would like to ask @RSNara help. I don't know very well [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java).

Additionally I wanted to point you out to the + button in the comment section, which allows you to propose a code changes to this pull request, or if you want you can write a pull request for my branch [fabriziobertoglio/react-native](https://github.com/fabriziobertoglio1987/react-native/tree/fix-dotted-borderLeft)

![image](https://user-images.githubusercontent.com/24992535/87010818-6bd51200-c1c7-11ea-9d51-3881ec54ce22.png)

**POINT 1**
>It is generally recommended against creating new instances in functions called in onDraw

You are right, makes perfect sense what you are saying, as probably onDraw is called many times and creating a new instance every time may have an effect on memory ([garbage collection](https://stackify.com/what-is-java-garbage-collection/))


My superficial understanding is that the existing/previous logic was built by creating new instances of `Path` in `onDraw`, my **ambitions** with this Pull Request was providing a **QuickFix**.

https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L128-L136

https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L488

**POINT 2**
>You may want to consider creating the Path where mPaint is created and reset it before using it here.

About this, `mPaint` is the color, `mPath` is the shape to be drawn. 

You propose to move the local variable `path` to a new instance variable of [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java) `mPathSingleBorder`.

Instead of storing the drawn `Path` value in 4 separate local variables named `pathTop`, `pathLeft`, `pathBottom` and `pathRight`, you propose to store the value in a `mPath` instance variable and [reset](https://developer.android.com/reference/kotlin/android/graphics/Path#reset()) it after `draw`. 


https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L1113-L1119

I allways prefer recursion over 1 instance variable, instead of writing 4 subroutines with 4 instance variables (like in this pull request).. It would make especially sense if we consider mantaining [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java) in the future and future developers may make use of the better code.

I don't believe [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java) will last very long and I believe it will be replaced from a new functionality.. I will add a new commit soon fixing this, but currently I am working on other pull request that fix functionality and parts of react-native which may give me a more relevant role in the project rather then [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java) (I say this because Facebook closed all those issues so maybe they plan to replace `ReactViewBackgroundDrawable` with `Fabric`).

My experience is that sometimes rewriting an entire functionality from scratch takes less time and brings better results then mantaining the current functionality, so I agree with Facebook approach with Fabric, but I also consider that I am doing this for **free** and may not be able to allways provide this services.

@benoitdion @RSNara if you are interested, I offer freelance consulting services at a very competitive rate of 10 USD per hour, if you know somebody interested, please forward my [Resume](https://fabriziobertoglio.xyz/). I am not proud of this code, all my [Freelance Clients](https://portfoliofabrizio.s3.eu-central-1.amazonaws.com/certificates/upwork.pdf) received much better services as I do make a great difference between free and paid service. Additionally I allways agree with clients, does not matter if they are right or wrong.

I'll add a fix to this soon, once I have more time. 

Thanks a lot for your feedback
I wish you a good day
Best Regards
Fabrizio Bertoglio

Reply:
I am just putting this pull request out there as as Quickfix for https://github.com/facebook/react-native/issues/24224 and 25+ issues with this bug https://github.com/facebook/react-native/issues/24224 https://github.com/facebook/react-native/issues/28695 https://github.com/facebook/react-native/issues/23651 https://github.com/facebook/react-native/issues/23475 https://github.com/facebook/react-native/issues/22256 https://github.com/facebook/react-native/issues/22226 https://github.com/facebook/react-native/issues/19234 https://github.com/facebook/react-native/issues/18285 https://github.com/facebook/react-native/issues/17344 https://github.com/facebook/react-native/issues/17343 https://github.com/facebook/react-native/issues/17251 https://github.com/facebook/react-native/issues/12817 https://github.com/facebook/react-native/issues/12403 https://github.com/facebook/react-native/issues/11042 https://github.com/facebook/react-native/issues/9343 https://github.com/facebook/react-native/issues/8236 https://github.com/facebook/react-native/issues/8105 https://github.com/facebook/react-native/issues/7838 https://github.com/facebook/react-native/issues/6721 https://github.com/facebook/react-native/issues/5411 https://github.com/facebook/react-native/issues/3159 https://github.com/facebook/react-native/issues/2335 https://github.com/facebook/react-native/issues/840 https://github.com/facebook/react-native/issues/27133 https://github.com/facebook/react-native/issues/28695 

All those issues have been closed, but actually never fixed. 

Additionally I wanted to point you out to the + button in the comment section, which allows you to propose a code changes to this pull request, or if you want you can write a pull request for my branch [fabriziobertoglio/react-native](https://github.com/fabriziobertoglio1987/react-native/tree/fix-dotted-borderLeft)

![image](https://user-images.githubusercontent.com/24992535/87010818-6bd51200-c1c7-11ea-9d51-3881ec54ce22.png)

**POINT 1**
>It is generally recommended against creating new instances in functions called in onDraw

You are right, makes perfect sense what you are saying, as probably onDraw is called many times and creating a new instance every time may have an effect on memory ([garbage collection](https://stackify.com/what-is-java-garbage-collection/))


My superficial understanding is that the existing/previous logic was built by creating new instances of `Path` in `onDraw`, my **ambitions** with this Pull Request was providing a **QuickFix**.

https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L128-L136

https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L488

**POINT 2**
>You may want to consider creating the Path where mPaint is created and reset it before using it here.

About this, `mPaint` is the color, `mPath` is the shape to be drawn. 

You propose to move the local variable `path` to a new instance variable of [ReactViewBackgroundDrawable](https://github.com/facebook/react-native/blob/03489539146556ec5ba6ba07ac338ce200f5b0f4/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java) `mPathSingleBorder`.

Instead of storing the drawn `Path` value in 4 separate local variables named `pathTop`, `pathLeft`, `pathBottom` and `pathRight`, you propose to store the value in a `mPath` instance variable and [reset](https://developer.android.com/reference/kotlin/android/graphics/Path#reset()) it after `draw`. 

I allways prefer recursion over 1 instance variable, instead of writing 4 subroutines with 4 instance variables (like in this pull request)

https://github.com/facebook/react-native/blob/40e940ad87c6d2ff70e76807c9ead85d158d2a1b/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L1113-L1119

I allways prefer recursion over 1 instance variable, instead of writing 4 subroutines with 4 instance variables (like in this pull request).

I'll add a fix once I have more time. 

Thanks a lot for your feedback
I wish you a good day
Best Regards
Fabrizio Bertoglio

Reply:
@benoitdion I added the changes you required. I don't believe this changes improve the pull request https://github.com/facebook/react-native/pull/29099/commits/cbc9f4503e580da15461dbaae5317a94f9b4e77a because I would rather have 4 local variables then 1 instance variable.

Applying recursion to this logic to avoid 4 if statements and instead grouping that logic in 1 function, requires a re-writing the exisiting logic included `drawRectangularBackgroundWithBorders` (extensive work):

1) Converting other local variables (left, top..etc.) to instance variables
2) Grouping the existing logic in methods that hold a single reponsability (setBorderLeft, setPath, setPaint...) 

I did not write that logic, I can apply temporary fixes, but my policy with open-source is to not re-write other developers code. I just write fixes for functionalities that do not work. 

I believe the code in this pull request is fine considering the design of the Class.

https://github.com/facebook/react-native/blob/cbc9f4503e580da15461dbaae5317a94f9b4e77a/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L1025-L1218

Please review this code and let me know how you would like to improve it.

Thanks a lot
Best Regards
Fabrizio Bertoglio

Reply:
@benoitdion I will try to implement your suggestions and improve this solution. Thanks

Reply:
Thanks a lot @benoitdion
>It is generally recommended against creating new instances in functions called in onDraw. You may want to consider creating the Path where mPaint is created and reset it before using it here.

I just noticed that the changes required in your message are included in commit https://github.com/facebook/react-native/pull/29099/commits/cbc9f4503e580da15461dbaae5317a94f9b4e77a 

1) > creating the Path where mPaint is created 

The Path `mPathForSingleBorder` is created where `mPaint` is created

https://github.com/facebook/react-native/blob/cbc9f4503e580da15461dbaae5317a94f9b4e77a/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L92
https://github.com/facebook/react-native/blob/cbc9f4503e580da15461dbaae5317a94f9b4e77a/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L106

2) >and reset it before using it here.

`mPathForSingleBorder` is resetted before drawing the border

https://github.com/facebook/react-native/blob/cbc9f4503e580da15461dbaae5317a94f9b4e77a/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java#L1114

3) I tried to replace `moveTo` and `lineTo` with `drawLine`, but the `dotted` style is not applied
4) I tried moving this logic to a separate function as displayed in https://github.com/fabriziobertoglio1987/react-native/commit/ed3230c983211da0351e576894926551871e5722 and then apply recursion using [invoke](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Method.html#invoke-java.lang.Object-java.lang.Object...-) and [lambda](https://www.w3schools.com/java/java_lambda.asp). 

I decided to not commit those changes as they do not make the code more readable, more clear and easier to search. Moving the logic to a separate function requires other developers to read/memorize an additional function.

The only reason to have a `drawSide` function below would be:

1) It would be easier to write a unit test. I would unit test this function with a mocking library and I would invoke the function `drawSide(width, start, end, spy)` and set an expectation on the `spy` canvas parameter.

I would expect that `spy` received method `:drawPath` with parameters `mPathForSingleBorder` and `mPaint` ([test-spy](https://github.com/rspec/rspec-mocks#test-spies)). I would verify `mPathForSingleBorder` and that `mPaint` is dotted.

```java
  private void drawSide(int width, PointF start, PointF end, Canvas canvas) {
    mPathForSingleBorder.reset();
    updatePathEffect(width);
    mPaint.setStrokeWidth(width);
    mPathForSingleBorder.moveTo(start.x, start.y);
    mPathForSingleBorder.lineTo(end.x, end.y);
    canvas.drawPath(mPathForSingleBorder, mPaint);
  }
```

2)  The below logic is repeated 4 times (`left`, `top`, `right`, `bottom`). The below code repetition could be avoided with a fifth parameter in `drawSide`. The fifth parameter would be a [`lambda`](https://www.w3schools.com/java/java_lambda.asp), the lambda would use [`invoke`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Method.html#invoke-java.lang.Object-java.lang.Object...-) to send a message (`:left`, `:top`, `:right`, `:bottom`) to `borderWidth` and `return` from the `drawSide` method if the result is 0. The same could be done to calculate `start` and `end` Point.  

```java
          if (borderWidth.left > 0) {
            int width = Math.round(borderWidth.left);
            PointF start = new PointF(left, top - borderWidth.top/2);
            PointF end = new PointF(left, bottom + borderWidth.bottom/2);
            drawSide(width, start, end, canvas);
          }
```

But I would not use reflection with this class as it would make the code less readable and harder to search.
I would prefer to write those changes in a separate Pull Request and give clear explanation of those changes.

I also believe that I should focus on fixing existing issues in react-native, for this reason I did not persist on the above changes.

This afternoon I will test further this functionality to detect any other potential issues
I remain available for further changes
Thanks a lot
Fabrizio Bertoglio